ballerina/dispatcher_service.bal

```diff
@@ -0,0 +1,324 @@
+                body: {
+                    jsonrpc: JSONRPC_VERSION,
+                    id: request.id,
+                    result: listToolsResult.cloneReadOnly()
+                }
+            };
+        }
+    }
+
+    private isolated function handleCallToolRequest(JsonRpcRequest request) returns http:BadRequest|http:Ok {
+        lock {
+            // Check if initialized
+            if !self.isInitialized {
+                JsonRpcError jsonRpcError = self.createErrorResponse(request.id, INVALID_REQUEST,
+                    "Client must be initialized before making requests");
+                return <http:BadRequest>{
+                    body: jsonRpcError.cloneReadOnly()
+                };
+            }
+
+            // Extract and validate parameters
+            CallToolParams|error params = request.cloneReadOnly().params.ensureType(CallToolParams);
+            if params is error {
+                JsonRpcError jsonRpcError = self.createErrorResponse(request.id, INVALID_PARAMS,
+                    string `Invalid parameters: ${params.message()}`);
+                return <http:BadRequest>{
+                    body: jsonRpcError.cloneReadOnly()
+                };
+            }
+
+            CallToolResult|error callToolResult = self.executeOnCallTool(params);
+            if callToolResult is error {
+                JsonRpcError jsonRpcError = self.createErrorResponse(request.id, INTERNAL_ERROR,
+                    string `Failed to call tool '${params.name}': ${callToolResult.message()}`);
+                return <http:BadRequest>{
+                    body: jsonRpcError.cloneReadOnly()
+                };
+            }
+
+            return <http:Ok>{
+                headers: {
+                    [SESSION_ID_HEADER]: self.sessionId ?: ""
+                },
+                body: {
+                    jsonrpc: JSONRPC_VERSION,
+                    id: request.id,
+                    result: callToolResult.cloneReadOnly()
+                }
+            };
+        }
+    }
+
+    private isolated function selectProtocolVersion(string requestedVersion) returns string {
+        foreach string supportedVersion in SUPPORTED_PROTOCOL_VERSIONS {
+            if supportedVersion == requestedVersion {
+                return requestedVersion;
+            }
+        }
+        return LATEST_PROTOCOL_VERSION;
+    }
+
+    private isolated function createErrorResponse(RequestId? id, int code, string message) returns JsonRpcError {
+        return {
+            jsonrpc: JSONRPC_VERSION,
+            id: id,
+            'error: {
+                code: code,
+                message: message
+            }
+        };
+    }
+
+    private isolated function executeOnListTools() returns ListToolsResult|error {
+        lock {
+            Service|AdvancedService? mcpService = self.mcpService;
+            if mcpService is AdvancedService {
+                return check invokeOnListTools(mcpService);
```

Comment:
```suggestion
                return invokeOnListTools(mcpService);
```

Reply:
If the method’s return type includes error, we can return directly without using check. It seems this pattern is repeated in several other places; please review and update.